pkg/resource/edit/operations_test.go

```diff
@@ -158,6 +158,71 @@ func TestDeletingDuplicateURNs(t *testing.T) {
 	})
 }
 
+func TestDeletingDuplicateProviderURN(t *testing.T) {
+	t.Parallel()
+
+	// Create duplicate provider resources
+	pA0 := NewProviderResource("a", "p1", "0")
+	pA1 := NewProviderResource("a", "p1", "1")
+
+	// Create a resource that depends on the duplicate Provider.
+	b1 := NewResource("b", pA0)
+	b2 := NewResource("b", pA1)
+
+	c := NewResource("c", pA1, b1.URN)
+
+	t.Run("do-target-dependents", func(t *testing.T) {
+		t.Parallel()
+		snap := NewSnapshot([]*resource.State{
+			pA0, pA1, b1, b2, c,
+		})
+
+		err := DeleteResource(snap, pA0, nil, true /* targetDependents */)
+		require.NoError(t, err)
+
+		// assert.Equals does a deep equals with the expected list.
+		assert.Equal(t, []*resource.State{
+			pA1, b2, c,
+		}, snap.Resources)
+	})
+
+	t.Run("do-not-target-dependents", func(t *testing.T) {
+		t.Parallel()
+		snap := NewSnapshot([]*resource.State{
+			pA0, pA1, b1, b2, c,
+		})
+
+		err := DeleteResource(snap, pA0, nil, false /* targetDependents */)
+		require.ErrorContains(t, err,
+			"Can't delete resource \"urn:pulumi:test::test::pulumi:providers:a::p1\" due to dependent resources")
+	})
+
+	t.Run("do-target-dependents-one-intermediate", func(t *testing.T) {
+		t.Parallel()
+		snap := NewSnapshot([]*resource.State{
+			pA0, pA1, b1, c,
+		})
+
+		err := DeleteResource(snap, pA0, nil, true /* targetDependents */)
+		require.NoError(t, err)
+		// assert.Equals does a deep equals with the expected list.
```

Comment:
You don't this comment on each Equals, we should all know Equal is a deep equal check.

Reply:
copy pasta from a previous test, I'll remove it everywhere.